### PR TITLE
Add `bee-trace` crate

### PR DESCRIPTION
**Pull request by Adam-Gleave**
Adds the new `bee-trace` crate, providing functionality that should be useful for diagnostics.

## `bee-trace`

The crate consists of a few components:
 - The `Observe` trait
 - A `tracing::Subscriber` implementation, as well as useful `Layer`s.
 - Diagnostic utilities (currently just one, in `Flamegrapher`).
 - `#[observe]` attribute procedural macro.

## The `Observe` trait

This trait provides the ability to wrap futures with an instrumentation span, similar to the `tracing_futures::Instrument` trait. This span has some default properties that make it easy to register and filter in a subscriber:

```rust
tracing::trace_span!(
    target: "bee::observe",
    "observed",
    observed.name = <my_function_name>,
    loc.file = <caller_file>,
    loc.line = <caller_line>,
    loc.col = <caller_column>,
);
```

Crucially, this span contains identical fields to those generated internally by `tokio` for task spawns (when compiled with `--cfg tokio_unstable`, which means that subscribers can interact with spans generated by `Observe` in the same way that they can with `tokio`s internal spans.

## Subscriber implementation

`bee-trace` provides a subscriber and two `Layer`s for dealing with spans and events:
 1) **Flamegraph** layer, produces a folded stack file detailing instrumented span stacks (either instrumented internally by `tokio` or by `bee-trace`) that can be used to generate a flamegraph of all *observed* code.
 2) **Logging** layer, allows `log` records to be converted into `tracing` events, and recreates full logging functionality within the subscriber itself, since it is impossible to set two loggers/subscribers at once. This means that logging remains consistent whether using `bee-trace` or not.

The subscriber can be initialised through a builder that can either set the global subscriber or return a `Layered` instance that can be further extended with more `Layer`s.

### Setting as the default subscriber:

```rust
// `Flamegrapher` handle returned, for building a flamegraph at the end of the run.
let flamegrapher = bee_trace::subscriber::build()
    .with_flamegraph_layer(stack_filename)
    .with_log_layer(log_config)
    .init()?
    .unwrap();
```

### Returning a `Layered` struct:

```rust
let (subscriber, flamegrapher) = bee_trace::subscriber::build()
    .with_flamegraph_layer(stack_filename)
    .with_log_layer(log_config)
    .finish()?;

// Extend the subscriber with external layers.
let subscriber = subscriber.with(console_layer);

// Set the global subscriber.
subscriber.init();
```

## `Flamegrapher`

Produces a flamegraph using the given folded stack file, using the `inferno` crate.

```rust
let flamegrapher = bee_trace::subscriber::build()
    .with_flamegraph_layer(stack_filename)
    .init()?
    .unwrap();

// Run some instrumented code...
// ...
// ...

// Programatically create the flamegraph file.
flamegrapher
    .with_graph_file("flamegraph.svg")?
    .write_flamegraph()?;
```

## `#[observe]` attribute

`bee-trace` provides a simple attribute proc-macro for instrumenting functions and futures with a span that specifies the `bee::observe` target. The location fields in the span will describe the location of the tagged function or future, and the `observed.name` field will specify the function name:

```rust
use bee_trace::observe;

#[observe]
pub async fn say_hello() {
    println!("hello");
}
```

This is equivalent to the following:

```rust
bee_trace::Observe(say_hello(), "say_hello").await;
```

This macro can be used with regular, non-`async` functions too, unlike the `Observe` trait.

## Examples

There is an example for each layer in `bee-trace/examples`. The flamegraph example produces this interactive graph:

example

# More stuff to add

After the initial work is finished, it makes sense to also:
 - Move the console subscriber code in `bee-common` to this crate. It makes more sense here, and should clean up `bee-common` a little bit. The feature flag would move from that crate to this one.
 - Add some customisation options for the flamegraph (module paths, function names rather than locations, group stacks by thread, etc).

There are also plenty of things we can do with some additional subscriber `Layer`s if we ever wanted more diagnostic tools. In addition to being able to write our own, there are existing crates that provide layers for Chrome tracing and the Tracy profiler, and more.

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

All existing tests have been run and have passed, a couple of new tests have been added for the procedural macro, and two examples are included in the `bee-trace` crate that can be run to verify that the subscriber and layers are working correctly.

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

File: bee-common/bee-common/src/logger/config.rs [hunk 1]
```diff
@@ -57,6 +57,13 @@ impl LoggerOutputConfigBuilder {
         self
     }
 
+    /// Sets a collection of exclusions of a logger output.
+    /// A message is logged only if one of the exclusions is *not* part of the log's metadata target.
+    pub fn target_exclusions(mut self, target_exclusions: &[&str]) -> Self {
+        self.target_exclusions = Some(target_exclusions.iter().map(|f| f.to_string()).collect::<Vec<String>>());
```
Comment (pvdrz) [inline bee-common/bee-common/src/logger/config.rs]: I think you should avoid the closure (clippy will complain about it iirc)

File: bee-common/bee-common/src/logger/config.rs [hunk 2]
```diff
@@ -91,6 +98,28 @@ pub struct LoggerOutputConfig {
...
+        &self.name
+    }
+
+    /// Returns the log level of the output.
+    pub fn level_filter(&self) -> LevelFilter {
+        self.level_filter
+    }
+
+    /// Returns the target filters of the output.
+    pub fn target_filters(&self) -> &Vec<String> {
```
Comment (pvdrz) [inline bee-common/bee-common/src/logger/config.rs]: Can you use `&[String]` to avoid the pointer indirection or is something else using a specific method of `Vec`?

File: bee-common/bee-common/src/logger/config.rs [hunk 3]
```diff
@@ -91,6 +98,28 @@ pub struct LoggerOutputConfig {
...
+        &self.target_filters
+    }
+
+    /// Returns the target exclusions of the output.
+    pub fn target_exclusions(&self) -> &Vec<String> {
```
Comment (pvdrz) [inline bee-common/bee-common/src/logger/config.rs]: Same here.

File: bee-trace/bee-trace-attributes/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        };
+
+        quote! {
+            bee_trace::Observe::observe(#observed_future, #fn_name).await
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (pvdrz) [inline bee-trace/bee-trace-attributes/src/lib.rs:157]: maybe you can use the same test suite setup as `bee-packable-derive` so we can keep tests in separated files. But I'd do that in a different PR. Not sure about the `expected_tokens` part tho.

File: bee-trace/bee-trace/examples/flamegraph.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        .with_flamegraph_layer(stack_filename)
+        .init()?
+        .unwrap();
+
+    println!("Running tasks for 5 seconds...");
+    let task1 = task::spawn(timeout(Duration::from_secs(5), run_tasks(500)));
+    let task2 = task::spawn(timeout(Duration::from_secs(5), run_tasks(200)));
+    let _ = tokio::join!(task1, task2);
+
+    let graph_filename = examples_dir.clone().join("flamegraph");
```
Comment (pvdrz) [inline bee-trace/bee-trace/examples/flamegraph.rs]: I think you can avoid this `.clone()` as `examples_dir` is not used anywhere else after this.

File: bee-trace/bee-trace/examples/flamegraph.rs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+    println!("Creating flamegraph at {}.svg", graph_filename.to_string_lossy());
```
Comment (pvdrz) [inline bee-trace/bee-trace/examples/flamegraph.rs]: I tend to favor `graph_filename.display()`

File: bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+/// to instrument [`task`](tokio::task)s, when compiling with `--cfg tokio_unstable` flags.
+pub struct FlamegraphLayer {
+    callsites: RwLock<HashSet<callsite::Identifier>>,
+    span_locations: RwLock<HashMap<span::Id, Option<SpanLocation>>>,
+    out_file: RwLock<BufWriter<File>>,
+}
+
+impl FlamegraphLayer {
+    pub(crate) fn new<P: AsRef<Path>>(stack_filename: P) -> Result<(Self, Flamegrapher), Error> {
+        let _ = *START;
```
Comment (pvdrz) [inline bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs:49]: Why do you need to deref this?

File: bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+            }
+            (_, _) => {}
+        }
+
+        // Ensure any downstream layers have access to all spans and events.
+        subscriber::Interest::always()
+    }
+
+    fn new_span(&self, attrs: &span::Attributes<'_>, id: &span::Id, _ctx: Context<'_, S>) {
+        if !self.is_tracked_callsite(&attrs.metadata().callsite()) {
```
Comment (pvdrz) [inline bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs]: can't this be:
```rust
       if self.is_tracked_callsite(&attrs.metadata().callsite()) {
           let location = SpanLocation::from_attributes(attrs);
           self.span_locations.write().insert(id.clone(), location);
       }
```

File: bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+            return;
+        }
+
+        let location = SpanLocation::from_attributes(attrs);
+        self.span_locations.write().insert(id.clone(), location);
+    }
+
+    fn on_enter(&self, id: &span::Id, ctx: Context<'_, S>) {
```
Comment (pvdrz) [inline bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs:173]: same question for all the other `on_/w+` methods

File: bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+impl TryFrom<LocationVisitor> for SpanLocation {
+    // This is a bit odd, but since a `LocationVisitor` can fail to record any location information,
+    // constructing a `SpanLocation` from one can fail. We don't need a specific error though, because
+    // `SpanLocation::from_attributes` will just return `None` instead.
+    type Error = ();
+
+    fn try_from(visitor: LocationVisitor) -> Result<Self, Self::Error> {
+        let LocationVisitor { file, line } = visitor;
+
+        file.zip(line).map(|(file, line)| SpanLocation { file, line }).ok_or(())
```
Comment (pvdrz) [inline bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs:222]: how come `zip_with` doesn't exists >:(

File: bee-trace/bee-trace/src/subscriber/layer/log.rs [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+/// [`collect_logs`](crate::subscriber::collect_logs) is called, or a [`LogTracer`](tracing_log::LogTracer)
+/// is initialised.
+pub struct LogLayer {
+    make_writers: Vec<LogTargetMakeWriter>,
+    fmt_events: LogFormatter,
+}
+
+impl<S> Layer<S> for LogLayer
+where
+    S: Subscriber + for<'a> LookupSpan<'a>,
```
Comment (pvdrz) [inline bee-trace/bee-trace/src/subscriber/layer/log.rs:122]: Doing fancy HKT bounds and all :trollface:

File: bee-trace/bee-trace/src/subscriber/layer/log.rs [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+                    Self::STDOUT_NAME => LogDest::Stdout,
+                    name => {
+                        let file = OpenOptions::new().write(true).create(true).append(true).open(name)?;
+                        LogDest::File(Mutex::new(file))
+                    }
+                };
+
+                Ok(LogTargetMakeWriter::new(LogTarget { filter: targets, dest }))
+            })
+            .collect::<Result<_, _>>()
```
Comment (pvdrz) [inline bee-trace/bee-trace/src/subscriber/layer/log.rs]: Can you add the `io::Error` to the collect's turbofish so you don't have to do it in the closure?

File: bee-trace/bee-trace/src/subscriber/mod.rs [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+            .map(layer::log_layer)
+            .map_or(Ok(None), |res| res.map(Some))
+    }
+
+    fn build_flamegraph_layer(
+        &mut self,
+    ) -> Result<(Option<layer::FlamegraphFilteredLayer>, Option<Flamegrapher>), Error> {
+        self.flamegraph_stack_file
+            .take()
+            .map(|stack_filename| layer::flamegraph_layer(stack_filename))
```
Comment (pvdrz) [inline bee-trace/bee-trace/src/subscriber/mod.rs]: Would it be too bad to put the contents of this map inside the `map_or` instead?

File: bee-common/bee-common/src/logger/config.rs (hunk 1 above)
Comment (Adam-Gleave) [reply bee-common/bee-common/src/logger/config.rs]: Ah, I didn't run clippy on bee-common! Good spot.

File: bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs (hunk 7 above)
Comment (Adam-Gleave) [reply bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs:49]: If it's not dereferenced here, it gets assigned the first time it is used elsewhere instead, which is inaccurate (and can cause issues with `duration_since`.

File: bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs (hunk 8 above)
Comment (Adam-Gleave) [reply bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs]: It can, but I prefer the early return way :trollface: Not that bothered about it though.

File: bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs (hunk 10 above)
Comment (Adam-Gleave) [reply bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs:222]: It does I think, but it's unstable 😢

File: bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs (hunk 7 above)
Comment (pvdrz) [reply bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs:49]: Oh I get it. I'd comment it just in case.

File: bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs (hunk 8 above)
Comment (pvdrz) [reply bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs]: @thibault-martinez we have an instance of the CAT theorem but with `A = Adam` fix plz :trollface:

File: bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs (hunk 10 above)
Comment (pvdrz) [reply bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs:222]: see you in 3 years :trollface:

File: bee-common/bee-common/src/logger/config.rs (hunk 2 above)
Comment (Adam-Gleave) [reply bee-common/bee-common/src/logger/config.rs]: Will do 👍

File: bee-trace/bee-trace/src/subscriber/layer/log.rs (hunk 12 above)
Comment (Adam-Gleave) [reply bee-trace/bee-trace/src/subscriber/layer/log.rs]: Didn't notice this -- yeah that would be better.

File: bee-trace/bee-trace/src/subscriber/layer/log.rs (hunk 11 above)
Comment (Adam-Gleave) [reply bee-trace/bee-trace/src/subscriber/layer/log.rs:122]: 😩

File: bee-trace/bee-trace-attributes/src/lib.rs (hunk 4 above)
Comment (Adam-Gleave) [reply bee-trace/bee-trace-attributes/src/lib.rs:157]: Sounds good, I can do another PR for that later.

**Comment by Adam-Gleave** [comment]
Looks like there's a problem with `tracing-futures` (it has a dependency on `tokio 0.1.22`, which has an issue). I'll try and fix it.

File: bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs (hunk 8 above)
Comment (thibault-martinez) [reply bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs]: I have a tendency of going for the least verbose option if it doesn't degrade readability, and I think here it wouldn't so I would agree with Christian

File: bee-trace/bee-trace-attributes/Cargo.toml [hunk 14]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "bee-trace-attributes"
+version = "0.1.0"
+edition = "2021"
+description = "Procedural macros for bee-trace"
+readme = "README.md"
+repository = "https://github.com/iotaledger/bee"
+license = "Apache-2.0"
+keywords = [ "iota", "tangle", "bee", "framework", "trace", "tracing", "attribute", "proc-macro" ]
```
Comment (thibault-martinez) [inline bee-trace/bee-trace-attributes/Cargo.toml]: 5 max allowed to publish to crates.io

File: bee-trace/bee-trace-attributes/Cargo.toml [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+[lib]
+proc-macro = true
+
+[dependencies]
+proc-macro2 = { version = "1.0.28", default-features = false }
+quote = { version = "1.0.9", default-features = false }
+syn = { version = "1.0.75", default-features = false, features = [ "full", "extra-traits", "parsing", "printing", "derive", "proc-macro", "clone-impls" ] }
+
+[dev-dependencies]
+bee-trace = { path = "../bee-trace" }
```
Comment (thibault-martinez) [inline bee-trace/bee-trace-attributes/Cargo.toml]: default-feature = false on dev deps as well, if it can help reduce CI time, why not...

File: bee-trace/bee-trace/CHANGELOG.md [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+### Removed
+
+### Fixed
+
+### Security -->
+
+## 0.1.0 - 2021-11-23
+
+### Added
```
Comment (thibault-martinez) [inline bee-trace/bee-trace/CHANGELOG.md:24]:
```suggestion
### Added

```

File: bee-trace/bee-trace/Cargo.toml [hunk 17]
```diff
@@ -0,0 +1,28 @@
+[package]
+name = "bee-trace"
+version = "0.1.0"
+authors = [ "IOTA Stiftung" ]
+edition = "2021"
+description = "Tracing and diagnostic tools for the bee node"
+readme = "README.md"
+repository = "https://github.com/iotaledger/bee"
+license = "Apache-2.0"
+keywords = [ "iota", "tangle", "bee", "framework", "trace", "tracing" ]
```
Comment (thibault-martinez) [inline bee-trace/bee-trace/Cargo.toml]: 5 max

File: bee-trace/bee-trace/Cargo.toml [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+lazy_static = "1.4.0"
+log = { version = "0.4.14", default-features = false }
+parking_lot = { version = "0.11.1", default-features = false }
+tracing = { version = "0.1.26", default-features = false }
+tracing-futures = { version = "0.2.5", default-features = false, features = [ "futures-01", "futures-03", "std", "std-future", "tokio" ] }
+tracing-log = { version = "0.1.2", default-features = false, features = [ "log-tracer", "std" ] }
+tracing-subscriber = { version = "0.2.20", default-features = false, features = [ "ansi", "fmt", "parking_lot", "registry" ] }
+
+[dev-dependencies]
+tokio = { version = "1.13", features = [ "full", "rt", "tracing" ] }
```
Comment (thibault-martinez) [inline bee-trace/bee-trace/Cargo.toml]: defaukt-features = false

File: bee-trace/bee-trace/Cargo.toml [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+description = "Tracing and diagnostic tools for the bee node"
+readme = "README.md"
+repository = "https://github.com/iotaledger/bee"
+license = "Apache-2.0"
+keywords = [ "iota", "tangle", "bee", "framework", "trace", "tracing" ]
+homepage = "https://www.iota.org"
+
+[dependencies]
+bee-common = { path = "../../bee-common/bee-common" }
+bee-trace-attributes = { path = "../bee-trace-attributes" }
```
Comment (thibault-martinez) [inline bee-trace/bee-trace/Cargo.toml]: default-features = false

File: bee-trace/bee-trace/README.md [hunk 20]
```diff
@@ -0,0 +1,3 @@
+# bee-trace
+
+Tracing and diagnostic tools for the bee node.
```
Comment (thibault-martinez) [inline bee-trace/bee-trace/README.md]:
```suggestion
Tracing and diagnostic tools for the bee framework.
```

File: bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs (hunk 8 above)
Comment (Adam-Gleave) [reply bee-trace/bee-trace/src/subscriber/layer/flamegraph.rs]: Sure, I'll change them 👍

**Comment by Adam-Gleave** [comment]
Looks like `inferno` uses the CDDL license. I think this is compatible with Apache?

**Review by pvdrz — DISMISSED** [review]

File: bee-common/bee-common/src/logger/config.rs [hunk 21]
```diff
@@ -53,7 +53,19 @@ impl LoggerOutputConfigBuilder {
     /// Sets a collection of filters of a logger output.
     /// A message is logged only if one of the filters is part of the log's metadata target.
     pub fn target_filters(mut self, target_filters: &[&str]) -> Self {
-        self.target_filters = Some(target_filters.iter().map(|f| f.to_string()).collect::<Vec<String>>());
+        self.target_filters = Some(target_filters.iter().map(ToString::to_string).collect::<Vec<String>>());
```
Comment (Alex6323) [inline bee-common/bee-common/src/logger/config.rs]: collect::<Vec<_>>()

File: bee-common/bee-common/src/logger/config.rs [hunk 22]
```diff
@@ -53,7 +53,19 @@ impl LoggerOutputConfigBuilder {
...
+    }
+
+    /// Sets a collection of exclusions of a logger output.
+    /// A message is logged only if one of the exclusions is *not* part of the log's metadata target.
+    pub fn target_exclusions(mut self, target_exclusions: &[&str]) -> Self {
+        self.target_exclusions = Some(
+            target_exclusions
+                .iter()
+                .map(ToString::to_string)
+                .collect::<Vec<String>>(),
```
Comment (Alex6323) [inline bee-common/bee-common/src/logger/config.rs]: same

File: bee-common/bee-common/src/logger/config.rs [hunk 23]
```diff
@@ -91,6 +103,28 @@ pub struct LoggerOutputConfig {
     pub(crate) target_exclusions: Vec<String>,
 }
 
+impl LoggerOutputConfig {
+    /// Returns the name of the output file, or `stdout` for standard output.
+    pub fn name(&self) -> &String {
```
Comment (Alex6323) [inline bee-common/bee-common/src/logger/config.rs]: return type &str ?

File: bee-common/bee-common/src/logger/config.rs [hunk 24]
```diff
@@ -91,6 +103,28 @@ pub struct LoggerOutputConfig {
...
+        &self.name
+    }
+
+    /// Returns the log level of the output.
+    pub fn level_filter(&self) -> LevelFilter {
+        self.level_filter
+    }
+
+    /// Returns the target filters of the output.
+    pub fn target_filters(&self) -> &[String] {
```
Comment (Alex6323) [inline bee-common/bee-common/src/logger/config.rs:124]: map to &[&str] ?

File: bee-common/bee-common/src/logger/config.rs [hunk 25]
```diff
@@ -91,6 +103,28 @@ pub struct LoggerOutputConfig {
...
+        &self.target_filters
+    }
+
+    /// Returns the target exclusions of the output.
+    pub fn target_exclusions(&self) -> &[String] {
```
Comment (Alex6323) [inline bee-common/bee-common/src/logger/config.rs:129]: same

File: bee-trace/bee-trace-attributes/src/lib.rs [hunk 26]
```diff
@@ -0,0 +1,224 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+//! Attribute macro for extending functions and futures with the `Observe` trait.
+
+#![deny(missing_docs, warnings)]
```
Comment (Alex6323) [inline bee-trace/bee-trace-attributes/src/lib.rs:6]: Sure you want to deny warnings? Forces us to quickly fix warnings on new compiler versions. We could be ahead of the users by fixing warnings already when they appear on beta, but could be easily forgotten I think.

File: bee-trace/bee-trace/Cargo.toml [hunk 27]
```diff
@@ -0,0 +1,28 @@
+[package]
+name = "bee-trace"
+version = "0.1.0"
+authors = [ "IOTA Stiftung" ]
+edition = "2021"
+description = "Tracing and diagnostic tools for the bee node"
```
Comment (Alex6323) [inline bee-trace/bee-trace/Cargo.toml]: I suggest writing: "Bee framework" instead of "Bee node" whenever possible (to emphasize that people could/should build their own nodes, in this case with the help of bee-trace)🙂

File: bee-trace/bee-trace/README.md [hunk 28]
```diff
@@ -0,0 +1,3 @@
+# bee-trace
+
+Tracing and diagnostic tools for the bee framework.
```
Comment (Alex6323) [inline bee-trace/bee-trace/README.md:3]: framework 🙂

**Comment by thibault-martinez** [comment]
> Looks like `inferno` uses the CDDL license. I think this is compatible with Apache?

You think or you're sure :trollface: ?

File: bee-trace/bee-trace-attributes/src/lib.rs (hunk 26 above)
Comment (Adam-Gleave) [reply bee-trace/bee-trace-attributes/src/lib.rs:6]: I forgot we stopped doing this! Will revert it 👍

File: bee-trace/bee-trace/README.md (hunk 28 above)
Comment (Adam-Gleave) [reply bee-trace/bee-trace/README.md:3]: :trollface:

File: bee-common/bee-common/src/logger/config.rs (hunk 24 above)
Comment (Adam-Gleave) [reply bee-common/bee-common/src/logger/config.rs:124]: Is there much to gain from the extra `.map( ... ).collect()`?

**Comment by Adam-Gleave** [comment]
> You think or you're sure :trollface: ?

I'm pretty sure, but I also am not a lawyer :trollface:

**Comment by Adam-Gleave** [comment]
Just a note that this PR doesn't contain any `console` stuff, that should be coming later today in another PR that builds on top of this one.

**Comment by Adam-Gleave** [comment]
Waiting on #952.

**Comment by Adam-Gleave** [comment]
Closing, since this has been moved to `common-rs` (iotaledger/common-rs#9).
